tap/src/dot11/nl.rs

```diff
@@ -74,178 +41,151 @@ pub enum InterfaceState {
+                        if_index = Some(attr.get_payload_as()?);
+                    },
+                    &Nl80211Attribute::IfName => {
+                        name = Some(attr.get_payload_as_with_len::<String>()?);
+                    },
+                    _ => {}
+                }
+            }
 
-        if attr.nla_type().nla_type().eq(&Nl80211Attribute::IfIndex) {
-            if_index = LittleEndian::read_u32(&attr.nla_payload().as_ref()[0..4]);
-        }
+            // we will always see a phy index, but we may not see the other attributes on the same message
+            if phy_index.is_some() {
+                phys_found.insert(phy_index.unwrap());
 
-        if attr.nla_type().nla_type().eq(&Nl80211Attribute::IfName) {
-            let mut bts = attr.nla_payload().as_ref().to_vec();
-            bts.pop();
-            name = String::from_utf8(bts).unwrap();
+                // check if we also saw the if_index on this message batch
+                if if_index.is_some() {
+                    if_indexes.insert(phy_index.unwrap(), if_index.unwrap());
+                }
+                // check if we also saw the name on this message batch
+                if name.is_some() {
+                    names_found.insert(phy_index.unwrap(), name.unwrap());
+                }
+            }
         }
     }
 
-    InterfaceResponse { name, phy_index, if_index }
-}
+    // now lets compile all the information we found regarding the interfaces, and if we have all attributes we will return them
+    let mut interfaces = HashMap::new();
 
-fn handle_phy_response(msg: Nlmsghdr<GenlId, Genlmsghdr<Nl80211Command, Nl80211Attribute>>) -> PhyResponse {
-    let mut phy_index: u32 = 0;
-    let mut supported_frequencies: Vec<u32> = Vec::new();
+    for phy_index in phys_found {
+        let name = names_found.get(&phy_index);
+        let if_index = if_indexes.get(&phy_index);
 
-    for attr in msg.get_payload().unwrap().attrs().iter() {
-        if attr.nla_type().nla_type().eq(&Nl80211Attribute::WiPhy) {
-            phy_index = LittleEndian::read_u32(&attr.nla_payload().as_ref()[0..4]);
+        // If any of the required attributes is missing, return an error.
+        if name.is_none() || if_index.is_none() {
+            return Err(DeError::new("Missing necessary attributes"));
         }
 
-        if attr.nla_type().nla_type().eq(&Nl80211Attribute::WiPhyBands) {
-            let payload = attr.nla_payload().as_ref();
-
-            let mut cursor: usize = 0;
-            loop {
-                if cursor+2 > payload.len() {
-                    break;
-                }
-
-                let element_length: usize = LittleEndian::read_u16(&payload[cursor..cursor+2]) as usize;
-                cursor += 2;
+        interfaces.insert(name.unwrap().clone(), Interface {
+            name: name.unwrap().clone(),
+            phy_index,
+            if_index: if_index.unwrap().clone(),
+        });
```

Comment:
There is a really good post about that here: https://blog.burntsushi.net/unwrap/

It is the same recommendation in effect today for the Rust std code and other core parts.

TLDR; It is ok to unwrap if panicking means there is a bug and is not expected behavior. 

I am ok though if we want to limit unwrap usage to tests only and try to match whenever possible and throw an error instead of panicing. Let's discuss this one so we can standardize one way or another.